lib/app/hive.dart

```diff
@@ -15,163 +16,217 @@ import 'package:schulcloud/news/news.dart';
+      HiveCache._createCollectionOfTypeId(
+        reader.readInt(),
+        reader.readString(),
+        reader.readStringList(),
+      );
+}
+
+/// A fetcher for an [IdCollection].
+class LazyIds<T extends Entity<T>> {
+  LazyIds({@required this.collectionId, @required this.fetcher});
+
+  final String collectionId;
+  // Id<IdCollection<T>> get _id => Id<IdCollection<T>>(collectionId);
+
+  final FutureOr<List<T>> Function() fetcher;
+
+  CacheController<List<T>> get controller {
+    return SimpleCacheController<List<T>>(
+      fetcher: fetcher,
+      loadFromCache: () async {
+        // TODO(marcelgarus): Activate the cache.
+        throw NotInCacheException();
+
+        // final ids = HiveCache.get(_id) ?? (throw NotInCacheException());
+        // return [
+        //   for (final itemId in ids.childrenIds)
+        //     HiveCache.get(itemId) ?? (throw NotInCacheException()),
+        // ];
+      },
+      saveToCache: (items) {
+        // TODO(marcelgarus): Activate the cache.
+        return;
+
+        // final collection = IdCollection<T>(
+        //   id: _id,
+        //   childrenIds: items.map((item) => item.id).toList(),
+        // );
+        // HiveCache.put<IdCollection<T>>(collection);
+        // items.forEach(HiveCache.put);
+      },
+    );
   }
 }
 
-class IdAdapter<T> extends TypeAdapter<Id<T>> {
-  IdAdapter(this.typeId);
+extension ListOfIds<T extends Entity<T>> on List<Id<T>> {
+  CacheController<List<T>> get controller {
+    return CacheController.combiningControllers(
+      map((id) => id.controller).toList(),
+    );
+  }
+}
 
-  @override
-  final int typeId;
+extension IdParsingList on List<dynamic> {
+  List<Id<T>> castIds<T extends Entity<T>>() {
+    return map((id) => Id<T>(id as String)).toList();
+  }
+}
 
-  @override
-  Id<T> read(BinaryReader reader) => Id<T>(reader.readString());
+// ignore: non_constant_identifier_names
+final HiveCache = HiveCacheImpl();
 
-  @override
-  void write(BinaryWriter writer, Id obj) => writer.writeString(obj.id);
+class Fetcher<T extends Entity<T>> {
+  Fetcher(this.fetch);
+
+  final Future<T> Function(Id<T> id) fetch;
+
+  Id<T> _createId(String id) => Id<T>(id);
+  IdCollection<T> _createCollection(String id, List<String> children) {
+    return IdCollection<T>(
+      id: Id<IdCollection<T>>(id),
+      childrenIds: children.map((child) => Id<T>(child)).toList(),
+    );
+  }
+}
+
+class HiveCacheImpl {
+  final _fetchers = <int, Fetcher>{};
+  Box<dynamic> _box;
+
+  Future<T> fetch<T extends Entity<T>>(Id<T> id) {
+    for (final fetcher in _fetchers.values) {
+      if (fetcher is Fetcher<T>) {
+        return fetcher.fetch(id);
+      }
+    }
+    throw UnsupportedError("We don't know how to fetch $T. Are you sure you "
+        'registered the type $T?');
+  }
+
+  Future<void> initialize() async => _box = await Hive.openBox('cache');
+
+  void registerEntityType<T extends Entity<T>>(
+      int typeId, Future<T> Function(Id<T> id) fetch) {
```

Comment:
```suggestion
      int typeId, FetchFunction<T> fetch) {
```
(See above comment for `FetchFunction`)